pkg/csi/manila/snapshot.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package manila
+
+import (
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/sharedfilesystems/v2/snapshots"
+	"k8s.io/klog"
+)
+
+const (
+	snapshotCreating      = "creating"
+	snapshotDeleting      = "deleting"
+	snapshotError         = "error"
+	snapshotErrorDeleting = "error_deleting"
+	snapshotAvailable     = "available"
+
+	snapshotDescription = "snapshotted-by=manila.csi.openstack.org"
+)
+
+// getOrCreateSnapshot retrieves an existing snapshot with name=snapID, or creates a new one if it doesn't exist yet.
+// Instead of waiting for the snapshot to become available (as getOrCreateShare does), CSI's ready_to_use flag is used to signal readiness
+func getOrCreateSnapshot(snapID snapshotID, sourceShareID string, manilaClient *gophercloud.ServiceClient) (*snapshots.Snapshot, error) {
+	var (
+		snapshot *snapshots.Snapshot
+		err      error
+	)
+
+	// First, check if the snapshot already exists or needs to be created
+
+	if snapshot, err = getSnapshotByName(string(snapID), manilaClient); err != nil {
+		if _, ok := err.(gophercloud.ErrResourceNotFound); ok {
+			// It doesn't exist, create it
+
+			req := snapshots.CreateOpts{
+				ShareID:     sourceShareID,
+				Name:        string(snapID),
+				Description: snapshotDescription,
+			}
+
+			var createErr error
+			if snapshot, createErr = snapshots.Create(manilaClient, req).Extract(); createErr != nil {
+				return nil, createErr
+			}
+
+		} else {
+			// Something else is wrong
+			return nil, fmt.Errorf("failed to probe for snapshot: %v", err)
+		}
+	} else {
+		klog.V(4).Infof("snapshot %s (snapshot ID %s) already exists", snapID, snapshot.ID)
```

Comment:
I think we should also check whether the sourceShareID is the same. Calling CreateSnapshot with the existing SnapshotID but different source is also an error.

Reply:
Good point, thanks!

Reply:
+1 if sourceShareID is different, we need to return codes.AlreadyExists error

Reply:
ID validation added in https://github.com/kubernetes/cloud-provider-openstack/pull/652/commits/a8bdd2356094f3b93d85eff20cf1eefc40a6d488